lib/sdm/container/3cache.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright © 2019 Atomist, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Atomist SDM Sample
+ * @description SDM that uses containers to build Node.js project and their Docker images.
+ * @tag sdm,container,node,docker,kaniko
+ * @instructions <p>This SDM is now listening for pushes to repos with
+ *               "package.json" files.  Push some new commits to one
+ *               such repo and this SDM will checkout the code, install
+ *               dependencies, run tests, and build a Docker image.
+ *               Make sure your Docker CLI is configured properly.</p>
+ */
+
+// atomist:code-snippet:start=containerCache
+/**
+ * This container goal example introduces input and output values for
+ * each goal using goal caching with the provided configuration.  The
+ * "docker" goal set is dependent on the "node" goal set.
+ */
+import { hasFile } from "@atomist/sdm";
+import {
+    CompressingGoalCache,
+    configure,
+    container,
+} from "@atomist/sdm-core";
+import * as os from "os";
+import * as path from "path";
+
+export const configuration = configure(async sdm => {
+    sdm.configuration.sdm.cache = {
+        enabled: true,
+        path: path.join(os.homedir(), ".atomist", "cache", "container"),
+        store: new CompressingGoalCache(),
+    };
+    return {
+        node: {
+            test: hasFile("package.json"),
+            goals: [
+                container("node", {
+                    containers: [{
+                        args: ["sh", "-c", "npm install && npm test"],
+                        env: [{ name: "NODE_ENV", value: "development" }],
+                        image: "node:12.4.0",
+                        name: "npm",
+                    }],
+                    output: [{
+                        classifier: "node-modules",
+                        pattern: { directory: "node_modules" },
```

Comment:
Would it be worth having an example on how to work with the output captured here? Maybe a simple example of log parsing or capturing mocha / maven test results?

Reply:
This output captures files on the file system to a persistent cache for the goal set, not log/stdout/stderr output. This output is used as input to the next goal. Is that sufficient or would you like an example that processes stdout/stderr from the container? Currently the stdout/stderr of the container is just sent to the goal's progress log.

Reply:
I spotted using the classifier to push data to the next goal, but I was thinking along the lines of accessing that output from within the SDM. Right now it looks like I can only get to that output data from another container task within the same goalset. I'm looking for something like a callback to be executed against the output of the container goal, which could modify/enrich the goal's result.

Reply:
Right now the log output is accessible from the `logInterpreter`. What are the kind of things you'd like to table to do when parsing the log output?

Reply:
> A simple example might something like: Parse the output of the test results upon goal completion, and add the number of tests passed/failed to the message of the goal result.

I think that might be possible today by using the `ProjectListener` stuff we got. I'll see if I can create an example for that.

Reply:
If you want to modify the value stored in the cache, you can create an intermediate goal that performs the modification and stores the updated value in the cache for the next goal.

If you want to parse/enrich the stdout/stderr, you can use the `logInterpreter` like @cdupuis said to get the content of the log and report on it via other means, e.g., a chat message, or provide an alternative `ProgressLogFactory` to the SDM that actually modifies the log in line.